http/middleware/external_dependency.go

```diff
@@ -0,0 +1,143 @@
+
+// ExternalDependencyHeaderName name of the HTTP header that is used for reporting
+const ExternalDependencyHeaderName = "External-Dependencies"
+
+// ExternalDependency middleware to report external dependencies
+func ExternalDependency(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var edc ExternalDependencyContext
+		edw := externalDependencyWriter{
+			ResponseWriter: w,
+			edc:            &edc,
+		}
+		r = r.WithContext(ContextWithExternalDependency(r.Context(), &edc))
+		next.ServeHTTP(&edw, r)
+	})
+}
+
+func AddExternalDependency(ctx context.Context, name string, dur time.Duration) {
+	ec := ExternalDependencyContextFromContext(ctx)
+	if ec == nil {
+		log.Ctx(ctx).Warn().Msgf("can't add external dependency %q with %s, because context is missing", name, dur)
+		return
+	}
+	ec.AddDependency(name, dur)
+}
+
+type externalDependencyWriter struct {
+	http.ResponseWriter
+	header bool
+	edc    *ExternalDependencyContext
+}
+
+// addHeader adds the external dependency header if not done already
+func (w *externalDependencyWriter) addHeader() {
+	if !w.header {
+		if len(w.edc.dependencies) > 0 {
+			w.ResponseWriter.Header().Add(ExternalDependencyHeaderName, w.edc.String())
+		}
+		w.header = true
+	}
+}
+
+func (w *externalDependencyWriter) Write(data []byte) (int, error) {
+	w.addHeader()
+	return w.ResponseWriter.Write(data)
+}
+
+func (w *externalDependencyWriter) WriteHeader(statusCode int) {
+	w.addHeader()
+	w.ResponseWriter.WriteHeader(statusCode)
+}
+
+// ContextWithExternalDependency creates a contex with the external provided dependencies
+func ContextWithExternalDependency(ctx context.Context, edc *ExternalDependencyContext) context.Context {
+	return context.WithValue(ctx, (*ExternalDependencyContext)(nil), edc)
+}
+
+// ExternalDependencyContextFromContext returns the external dependencies context or nil
+func ExternalDependencyContextFromContext(ctx context.Context) *ExternalDependencyContext {
+	if v := ctx.Value((*ExternalDependencyContext)(nil)); v != nil {
+		return v.(*ExternalDependencyContext)
+	}
+	return nil
+}
+
+// ExternalDependencyContext contains all dependencies that where seed
+// during the request livecycle
+type ExternalDependencyContext struct {
+	mu           sync.RWMutex
+	dependencies []externalDependency
+}
+
+func (c *ExternalDependencyContext) AddDependency(name string, duration time.Duration) {
+	c.mu.Lock()
+	c.dependencies = append(c.dependencies, externalDependency{
+		Name:     name,
+		Duration: duration,
+	})
+	c.mu.Unlock()
+}
+
+// String formats all external dependencies
+func (c *ExternalDependencyContext) String() string {
+	var buf bytes.Buffer
+	sep := len(c.dependencies) - 1
+	for _, dep := range c.dependencies {
+		buf.WriteString(dep.String())
+		if sep > 0 {
+			buf.WriteByte(',')
+			sep--
+		}
+	}
+	return buf.String()
+}
+
+// Parse a external dependency value
+func (c *ExternalDependencyContext) Parse(s string) {
```

Comment:
no error reporting/handling? How would we now that we have invalid data?

We simply assume we never right invalid data, right?

Reply:
Error handling doesn't make too much sense. The server response is either by the spec or not. So if not we drop the information otherwise we take it.